### ISSUE TITLE
Workflow questions (meta)

### ISSUE THREAD

**Issue by j-fu** [JuliaPackaging/Yggdrasil#608]
Hi,

I have some questions about the workflow with Yggdrasil.

Assume I develop package libA written in C and package A.jl  in Julia wrapping libA. So I want to provide libA_jll.jll to Yggdrasil.

- Can I create and test  libA_jll.jll locally and test its interplay with A.jl ?
- Same question for new versions of libA_jll.jll: how to test locally before submitting the PR ?

Possibly I can access the build in the BinaryBuilder workspace in the local filesystem. Is this the intended way ? Are there tools available to handle this ?

**Comment by giordano** [comment]
You can build locally the tarballs and publish the JLL package: if you call any `build_tarballs.jl` script with
```
$ julia build_tarballs.jl --help
```
you'll see
```
        --deploy=<repo>     Deploy binaries and JLL wrapper code to a github
                            release of an autogenerated repository.  Uses
                            `github.com/JuliaBinaryWrappers/<name>_jll.jl` by
                            default, unless `<repo>` is set, in which case it
                            should be set as `<owner>/<name>_jll.jl`.  Setting
                            this option is equivalent to setting `--deploy-bin`
                            and `--deploy-jll`.
```

**Comment by j-fu** [comment]
Hi thanks for the hint - it works, see  https://github.com/j-fu/VTKMinimal_jll.jl .
This is a minimal build of VTK without HDF5 and MPI, avoiding transient linking issues wrt. MPI but 
sufficient for visualization purposes.

**Comment by j-fu** [comment]
Oh and how to refere to this as a dependency in another build_tarballs.jl ?
If it would sit  on Yggdrasil I would write
`dependencies=[ "VTKMinimal_jll.jl"]`.
How I proceed in this case ? 

I just tried the following without success:
```
dependencies = [
    Dependency(Pkg.PackageSpec(name="VTKMinimal_jll",url="https://github.com/j-fu/VTKMinimal_jll.jl"))
]
```
(it throws `ERROR: LoadError: Invalid dependency specifications!`)

**Comment by giordano** [comment]
Can you please show the stacktrace?

**Comment by j-fu** [comment]
```
[ Info: Building for x86_64-linux-gnu-cxx11
  Updating registry at `~/.julia/registries/General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
┌ Warning: Unable to resolve VTKMinimal_jll
└ @ BinaryBuilder ~/.julia/packages/BinaryBuilder/uXcvN/src/Dependencies.jl:85
ERROR: LoadError: Invalid dependency specifications!
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] #autobuild#345(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Union{Nothing, String}, ::Bool, ::Bool, ::Any, ::Any, ::typeof(autobuild), ::AbstractString, ::AbstractString, ::VersionNumber, ::Array{#s993,1} where #s993<:BinaryBuilder.AbstractSource, ::AbstractString, ::Array{T,1} where T, ::Array{#s992,1} where #s992<:Product, ::Array{#s991,1} where #s991<:BinaryBuilder.AbstractDependency) at /home/fuhrmann/.julia/packages/BinaryBuilder/uXcvN/src/AutoBuild.jl:542
 [3] (::BinaryBuilder.var"#kw##autobuild")(::NamedTuple{(:verbose, :debug, :meta_json_stream),Tuple{Bool,Bool,Nothing}}, ::typeof(autobuild), ::String, ::String, ::VersionNumber, ::Array{GitSource,1}, ::String, ::Array{Linux,1}, ::Array{LibraryProduct,1}, ::Array{Dependency,1}) at ./none:0
 [4] #build_tarballs#323(::Any, ::typeof(build_tarballs), ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any) at /home/fuhrmann/.julia/packages/BinaryBuilder/uXcvN/src/AutoBuild.jl:193
 [5] build_tarballs(::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any) at /home/fuhrmann/.julia/packages/BinaryBuilder/uXcvN/src/AutoBuild.jl:25
 [6] top-level scope at /home/fuhrmann/Wias/work/julia/dev/Yggdrasil/V/vtkfig/build_tarballs.jl:47
 [7] include at ./boot.jl:328 [inlined]
 [8] include_relative(::Module, ::String) at ./loading.jl:1105
 [9] include(::Module, ::String) at ./Base.jl:31
 [10] exec_options(::Base.JLOptions) at ./client.jl:287
 [11] _start() at ./client.jl:460
in expression starting at /home/fuhrmann/Wias/work/julia/dev/Yggdrasil/V/vtkfig/build_tarballs.jl:47
```

**Comment by giordano** [comment]
That happens because [`resolve_jlls`](https://github.com/JuliaPackaging/BinaryBuilder.jl/blob/1a23569d0fb7abce7d6ed5b6fd02bccc93acb0fe/src/Dependencies.jl#L63-L90) looks up the packages in the registry.  A quick work around for you would be to add the package to your local registry (create a branch out of `master`)

**Comment by j-fu** [comment]
![vtkfig-2020-03-11T01:09:04](https://user-images.githubusercontent.com/9221937/76369941-8db3aa80-6335-11ea-8d69-da69758cb2b8.png)

**Comment by j-fu** [comment]
Ok i  set up a  clean registry by myself, anyway I wanted to check this out. 
In the moment the stuff is in
- https://github.com/j-fu/JFURegistry.jl
- https://github.com/j-fu/VTKMinimal_jll.jl
- https://github.com/j-fu/vtkfig_jll.jl
- https://github.com/j-fu/vtkfig
- https://github.com/j-fu/VTKView.jl

Will clean this a bit and send PRs in the next days and present this at vizcon2 if its not killed by the virus....
Thanks for all the hints!

**Comment by j-fu** [comment]
Just to keep being annoying: what are the chances for mac and win ?  I've seen that the X stuff so far seems to be linux bound. VTK surely builds on the mac and I've seen this running (including vtkfig) in the unix subsystem of win10.

**Comment by giordano** [comment]
Does this require X on macOS and Windows?

**Comment by giordano** [comment]
I don't see any dependency on X in [Homebrew](https://formulae.brew.sh/formula/vtk)

**Comment by j-fu** [comment]
They appear to use QT which may  work without. So that's another interesting question: how to dispatch among dependencies between operating systems. May be just by  having OS dependent dependency lists ? 

And Apple is moving the goal posts anyway: https://discourse.slicer.org/t/apple-to-deprecate-opengl-in-osx-10-14-mojave/3075 ...

**Comment by giordano** [comment]
If a JLL package doesn't have a tarball for an operating system, it will simply do nothing in that case, so it's safe to unconditionally depend on Xorg libraries

**Comment by j-fu** [comment]
If using the jll   I can check if libvtkfig is defined  and dispatch to  a custom build if not ?

**Comment by giordano** [comment]
Could you please elaborate a little bit more?

**Comment by j-fu** [comment]
From your post I understand that if the tarball for an OS is missing, the library products won't be available, and the global variables holding their names won't be  defined. 
So in principle in this case I can hint on the possibility to build the libraries locally and to put them in to whatever LD_LIBRARY_PATH on that system is so they can be found from ccall.

**Comment by j-fu** [comment]
vtk on mingw seems to be possible: https://packages.msys2.org/package/mingw-w64-x86_64-vtk?repo=mingw64 and they seem to be bases on QT as well.

... will be back tomorrow...

**Comment by giordano** [comment]
> From your post I understand that if the tarball for an OS is missing,

This is correct

> the library products won't be available, and the global variables holding their names won't be defined.

This is true if you're talking about the JLL package used in as normal package in Julia.  I thought we were talking about building within BinaryBuilder

It's in an early stage, but Qt is available in #586

**Comment by ViralBShah** [comment]
It seems that the original issue posted here has been discussed.
